### PR TITLE
[C/ObjC/C++/ObjC++] improved syntax, typedef detection improvement

### PR DESCRIPTION
**Pull request by dannftk**
`typedef`s of `struct|class|enum|union` are detected wiser, target defined types are recognized as types rather than nothing

Changes are applied to C, C++, ObjC, ObjC++ syntaxes

### CODE REVIEW

**Comment by dannftk** [comment]
Is anyone to react on the PR?

**Comment by rchl** [comment]
When the ST developer has time, he will come and merge multiple PRs in one swipe.
It's not uncommon for PRs to hang for months before that happens though so be patient.

In the meantime, if you want to use your changes, your best bet is too have local override for that syntax.

**Comment by dannftk** [comment]
How to request someone to run CI on the PR?

File: C++/syntax_test_c.c [hunk 1]
```diff
@@ -1028,3 +1029,11 @@ int control_keywords()
   return 123;
   /* <- keyword.control.flow.return */
 }
+
+typedef struct MyStructure {} MyStructure_t;
```
Comment (deathaxe) [inline C++/syntax_test_c.c:1033]: I'd try to keep test cases logically grouped. So moving this one before or after the one in line 290 would be suitable.

File: C++/syntax_test_cpp.cpp [hunk 2]
```diff
@@ -2753,3 +2753,11 @@ void sayHi()
 /**
       *
 /*    ^ comment.block.c punctuation.definition.comment.c */
+
+typedef struct MyStructure {} MyStructure_t;
```
Comment (deathaxe) [inline C++/syntax_test_cpp.cpp:2757]: The same here. There's a test case at line 410, which this one could moved to for better grouping. The current one at 410 could also be tweaked to test your change.

**Review by deathaxe — CHANGES_REQUESTED** [review]
The intention of my request is to keep fragmentation of test cases low. Maybe already too late in some syntaxes, I know.

Otherwise I find it a simple but effective improvement.

**Review by deathaxe — DISMISSED** [review]

**Comment by dannftk** [comment]
@deathaxe, I've done everything how you asked me to do

**Review by deathaxe — APPROVED** [review]

**Review by keith-hall — APPROVED** [review]

**Pull request merged**